src/Microsoft.DotNet.XHarness.CLI/Commands/WebServer.cs

```diff
@@ -0,0 +1,149 @@
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.Options;
+using Microsoft.DotNet.XHarness.CLI.CommandArguments;
+
+namespace Microsoft.DotNet.XHarness.CLI.Commands
+{
+    public class WebServer
+    {
+        internal static async Task<ServerURLs> Start(TestCommandArguments arguments, ILogger logger, Func<WebSocket, Task>? onConsoleConnected, CancellationToken token)
+        {
+            var host = new WebHostBuilder()
+                .UseKestrel()
+                .UseContentRoot(arguments.AppPackagePath)
+                .UseStartup<TestWebServerStartup>()
+                .ConfigureLogging(logging =>
+                {
+                    logging.AddConsole().AddFilter(null, LogLevel.Warning);
+                })
+                .ConfigureServices((ctx, services) =>
+                {
+                    services.AddRouting();
+                    services.AddSingleton<ILogger>(logger);
+                    services.Configure<TestWebServerOptions>(ctx.Configuration);
+                    services.Configure<TestWebServerOptions>(options =>
+                    {
+                        options.OnConsoleConnected = onConsoleConnected;
+                        foreach (var (middlewarePath, middlewareTypeName) in arguments.WebServerMiddlewarePathsAndTypes)
+                        {
+                            var extensionAssembly = AssemblyLoadContext.Default.LoadFromAssemblyPath(middlewarePath);
+                            var middlewareType = extensionAssembly?.GetTypes().Where(type => type.Name == middlewareTypeName).FirstOrDefault();
+                            if (middlewareType == null)
+                            {
+                                var message = $"Can't find {middlewareTypeName} middleware in {middlewarePath}";
+                                logger.LogError(message);
+                                throw new Exception(message);
+                            }
+                            options.EchoServerMiddlewares.Add(middlewareType);
+                        }
+                    });
+                })
+                .UseUrls("http://127.0.0.1:0", "https://127.0.0.1:0")
+                .Build();
+
+            await host.StartAsync(token);
+
+            var ipAddress = host.ServerFeatures
+                .Get<IServerAddressesFeature>()?
+                .Addresses
+                .Where(a => a.StartsWith("http:"))
+                .FirstOrDefault();
+
+            var ipAddressSecure = host.ServerFeatures
+                .Get<IServerAddressesFeature>()?
+                .Addresses
+                .Where(a => a.StartsWith("https:"))
+                .FirstOrDefault();
+
+            if (ipAddress == null || ipAddressSecure == null)
+            {
+                throw new InvalidOperationException("Failed to determine web server's IP address or port");
+            }
+
+            return new ServerURLs(ipAddress.Substring(ipAddress.LastIndexOf('/') + 1),
```

Comment:
What is `ipAddress.Substring(ipAddress.LastIndexOf('/') + 1` for?

Reply:
You can use `new Uri(ipAddress)`, and then use `$"{uri.Host}:{uri.Port}"`

Reply:
done